### PR TITLE
Improves bot performance

### PR DESCRIPTION
**Pull request by mako-taco**
- removes sleep from step-walker in favor of minimum-tick-time in main loop
- prevents evolve-all-worker from running every tick
- transferring pokemon worker runs every nth tick (configurable)
- adds performance time logging with --log_timing_info (defaults to false)

image

image

```
    "location": "Old Town Alexandria",
    "walk": 40.0,
    "minimum_tick_time": 2.0,
    "action_wait_min": 1,
    "action_wait_max": 4,
```

### CODE REVIEW

File: pokemongo_bot/tree_config_builder.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def build(self):
             task_config = task.get('config', {})
 
             worker = self._get_worker_by_name(task_type)
-            instance = worker(self.bot, task_config)
+            instance = worker(self.bot, task_config, task_type)
```
Comment (TheSavior) [inline pokemongo_bot/tree_config_builder.py]: Why do we want to pass in the name of the task to the task itself? It already knows. It's its class name

File: pokemongo_bot/cell_workers/transfer_pokemon.py [hunk 2]
```diff
@@ -5,7 +5,13 @@
 from pokemongo_bot.cell_workers.base_task import BaseTask
 
 class TransferPokemon(BaseTask):
+    def initialize(self):
```
Comment (TheSavior) [inline pokemongo_bot/cell_workers/transfer_pokemon.py]: I think we need a smarter way to speed up this task, I'm definitely a :-1: on checking the tick. I don't even like that we have a flag for the first tick.

Another approach here would be for the task to have a list of the pokemon it saw on the last tick, and then on the next tick if the pokemon have changed, run the transfer check on only the relevant ones.

I think this should be undone in this PR and done properly in a standalone PR.

File: pokemongo_bot/tree_config_builder.py (hunk 1 above)
Comment (mako-taco) [inline pokemongo_bot/tree_config_builder.py]: I don't know python

**Comment by TheSavior** [comment]
@mako-taco I think a lot of this functionality is valuable, but this PR is changing too many things. It makes it too hard for us to review, and revert individual pieces if necessary. It also keeps us from being able to make intelligent merge decisions to reduce the number of merge conflicts people will have.

Can you break this into individual PRs that can be reviewed separately? Logging, EvolveAll, TransferPokemon, StepWalker? Then we can review the individual merits and implementation of each thing. Doing this also means they will also get merged faster.

File: pokemongo_bot/cell_workers/transfer_pokemon.py (hunk 2 above)
Comment (mako-taco) [inline pokemongo_bot/cell_workers/transfer_pokemon.py]: I think it should only run when the pokemon bag is full or on startup, but I don't think we have a way to check number of pokemon w/o hitting the server, and this was a much faster fix.  Do you have any ideas?

**Comment by TCB13** [comment]
I wish this was added!

**Pull request closed without merging**